File: programmers/Level. 3/정수 삼각형.py
# 프로그래머스 Lv.3 정수 삼각형

def solution(triangle):
    # 삼각형의 위 층 숫자를 아래 층 숫자에 더한 뒤 위 층을 삭제하는 형태
    # 아래 층 숫자에 수를 더할 때, 위 층 중 왼쪽 / 오른쪽 더 큰 수를 더해야 함
    
    for i in range(1, len(triangle)) : # 1부터 삼각형 층수까지 반복
        for j in range(i + 1) : # 0부터 해당 층수의 배열 끝까지 반복
            if j == 0 : # 제일 왼쪽 숫자의 경우 바로 위 층의 가장 왼쪽 숫자를 더함
                triangle[i][j] += triangle[i - 1][0]
            elif j == i : # 제일 오른쪽 숫자의 경우 바로 위층의 가장 오른쪽 숫자를 더함
                triangle[i][j] += triangle[i - 1][-1]
            else : # 그 외의 모든 경우는 위 층의 숫자 두 개중 큰 수를 더함
                triangle[i][j] += max(triangle[i - 1][j - 1], triangle[i - 1][j])
                
    return max(triangle[-1])

File: programmers/Level. 3/정수 삼각형.cpp
// 프로그래머스 Lv.3 정수 삼각형

#include <string>
#include <vector>
#include <algorithm>

using namespace std;

int solution(vector<vector<int>> triangle) {
    // 1. 이중 반복문을 통해 위 층의 숫자를 아래층에 더함
    for(int i = 1; i<triangle.size(); i++)
        for(int j = 0; j<triangle[i].size(); j++)
        {
            // 가장 왼쪽 숫자라면 왼쪽 숫자를 더함
            if(j == 0)
                triangle[i][j] += triangle[i - 1].front();
            // 가장 오른쪽 숫자라면 오른쪽 숫자를 더함
            else if(j == triangle[i].size() - 1)
                triangle[i][j] += triangle[i - 1].back();
            // 그 외라면 위의 두 숫자중 큰 숫자를 더함
            else
            {
                int bigger = max(triangle[i - 1][j - 1],triangle[i - 1][j]);
                triangle[i][j] += bigger;
            }
        }
    
    // 2. 가장 마지막 층까지 도달했다면 *max_element로 최댓값 리턴
    return *max_element(triangle.back().begin(), triangle.back().end());
}